### ISSUE TITLE
Escopo 4 — Optimization 3, Part B: partial top-k selection instead of full vocabulary sort

### ISSUE THREAD

**Issue by fabious054** [fabious054/CandleCLI#3]
## Context

Optimization 3 (issue #1), Part A is merged — `lm_head` now runs through our own `Q8_0Matmul` kernel instead of candle's `QMatMul`. Live throughput went from 7.7 to 9.0/7.3/6.4 tok/s across three different prompts.

Part B targets a different waste, unrelated to the matmul kernel: `TopK::sample` (`src/sampler/top_k.rs`) sorts the **entire vocabulary** (~152k logits) just to keep the top `k` (usually 40):

```rust
let mut idx: Vec<usize> = (0..logits.len()).collect();
idx.sort_unstable_by(|&a, &b| {
    logits[b].partial_cmp(&logits[a]).unwrap_or(Ordering::Equal)
});
idx.truncate(self.k);
```

This is `O(n log n)` over 152k elements every single decode step, purely to select 40. A partial selection (`slice::select_nth_unstable_by`, Rust std) partitions around the k-th element in `O(n)` average time, without a full sort.

## What to implement

Replace the full sort in `TopK::sample` with `select_nth_unstable_by`:

1. Partition `idx` (or an index/value pair structure) around position `self.k` using descending order by logit value — everything before that position is guaranteed >= everything after, but *not* internally sorted.
2. Softmax needs the top-k *values* (order doesn't matter for correctness — softmax is order-independent), so no further sort is needed after the partition, unless there's a reason to keep deterministic ordering for something downstream (check `sample_multinomial` — it iterates in whatever order it receives, so partition order is fine).
3. Same result as today: same top-k set, same probabilities, same sampling — only *how* the top-k set is found changes.

## Acceptance criteria

- [ ] `TopK::sample` uses `select_nth_unstable_by` (or equivalent partial-selection approach), no full sort of the vocabulary
- [ ] Same top-k *set* selected as the current full-sort implementation (write a test comparing both approaches on the same synthetic logits — same set of top-k indices, regardless of internal order)
- [ ] Existing sampler tests still pass (`top_k_one_always_picks_the_max`, etc.)
- [ ] `cargo test --lib` green
- [ ] Manual validation by Fabio with the real model — `/top-k` still behaves the same from the user's perspective (same style of output, no crash, no behavior change)

## Out of scope

- `TopP::sample` — check whether it has the same full-sort issue while implementing this, but do not change it here without discussion first; if it does, that becomes a separate follow-up issue.
- Anything related to `lm_head` or the matmul kernels (Part A, already merged).

## References

- Trello: "Diretriz — Performance: máximo do hardware com mínimo de esforço, sempre"
- ADR-0003, ADR-0004 (not directly relevant here — no precision divergence expected, this is a pure algorithmic change with identical output)

**Comment by fabious054** [comment]
## Implemented — ready for local testing

Branch `escopo4/otimizacao-3-partb-topk-select`, commit `3668cf4`.

### What changed (`src/sampler/top_k.rs` only)

Extracted the selection logic into `top_k_indices(logits, k)`, which now uses `select_nth_unstable_by` instead of a full sort. Same top-k set, same softmax, same sampling — only the algorithm that finds the set changed (`O(n)` average vs `O(n log n)`).

New test `top_k_indices_matches_full_sort_reference_set` verifies the partial-selection result matches the old full-sort result as a *set* (order doesn't matter downstream) across k = 1, 5, 40, 200, 999 on a 1000-element synthetic array.

### `TopP` — checked, left untouched

`TopP::sample` also sorts the full vocabulary, but for a structural reason: it needs elements in probability order to find the cumulative-`p` cutoff, which is a data-dependent position, not a fixed `k`. `select_nth_unstable_by` doesn't apply the same way there. Documented in the commit message; a separate issue if this needs attention later.

### To run

```
git fetch origin
git checkout escopo4/otimizacao-3-partb-topk-select
git pull origin escopo4/otimizacao-3-partb-topk-select
cargo test --lib
```

No precision divergence expected here — this is a pure algorithmic change, same output. If any test fails, that would be a real bug, not a documented/expected divergence like Parts A's ADR-0004 risk.

After tests pass, `cargo run --release`, `/top-k <value>`, and confirm the sampling still feels the same from the user's side (no crash, same style of variation between runs).
